gui: Add GUIUtil::bringToFront

promag opened the pull request:
The sequence `show -> raise -> activateWindow` is factored out to the new function `GUIUtil::bringToFront` where a macOS fix is added in order to fix #13829.

Also included:
 - simplification of `BitcoinGUI::showNormalIfMinimized`
 - simplified some connections to `BitcoinGUI::showNormalIfMinimized`
 - added missing connections to `BitcoinGUI::showNormalIfMinimized`.

promag committed:
qt: Add GUIUtil::bringToFront

promag committed:
qt: Use GUIUtil::bringToFront where possible

promag committed:
qt: All tray menu actions call showNormalIfMinimized

promag commented:
@jonasschnelli @Sjors @ken2812221 review/test much appreciated.

ken2812221 commented on src/qt/bitcoingui.cpp:
```diff
@@ -274,15 +274,15 @@ void BitcoinGUI::createActions()
     // can be triggered from the tray menu, and need to show the GUI to be useful.
     connect(overviewAction, &QAction::triggered, this, static_cast<void (BitcoinGUI::*)()>(&BitcoinGUI::showNormalIfMinimized));
```
Any reason to keep this one?

promag replied:
Thanks.

laanwj commented:
I think it would make sense, for ease of review, to split off the fix to #13829 from pure refactoring.

promag commented:
Which refactor?

Note that the issue is observed in more places than the one reported in #13829.

Sjors commented on src/qt/guiutil.cpp:
```diff
@@ -357,6 +367,21 @@ bool isObscured(QWidget *w)
+{
+#ifdef Q_OS_MAC
+    id app = objc_msgSend((id) objc_getClass("NSApplication"), sel_registerName("sharedApplication"));
+    objc_msgSend(app, sel_getUid("activateIgnoringOtherApps:"), YES);
```
Apple [documentation](https://developer.apple.com/documentation/appkit/nsapplication/1428468-activateignoringotherapps?language=objc) suggests that this normally shouldn't be needed, which suggests there's a deeper problem somewhere.

fanquake replied:
I'd also like to know a bit more about this, and the new `<objc/objc-runtime.h>` #include.

promag replied:
Sure.

promag replied:
Added comment.

hebasto replied:
According to the `sel_getUid` [documentation](https://developer.apple.com/documentation/objectivec/1418625-sel_getuid?language=objc):
> The implementation of this method is identical to the implementation of `sel_registerName`.

On master(750415701):
- `sel_registerName` used 4 times
- `sel_getUid` not used at all

Could `sel_getUid` replace with `sel_registerName`?

Sjors reviewed (approved):
tACK eed869b on macOS, this indeed fixes #13829.

The change from `static_cast<void (BitcoinGUI::*)()>` to `[this]` is most eye pleasing :-)

I think @laanwj is referring to e.g. the new `bringToFront()` function which contains both existing and new code. I didn't find it too confusing though.

fanquake commented:
Testing current master on macOS 10.13.6 https://github.com/bitcoin/bitcoin/commit/2070a545e2c4e8c2dff4c0fb40b8fd2f68fc2ca.
If `Command + H` or the menu item is used to hide the window:
- Clicking the task bar icon shows the window.
- Clicking `show/hide` does not work.
- Clicking `show` (bottom of the menu ) does work.
- If `Send` or `Receive` is clicked, the window changes to the correct tab, but the window is not shown.
- `Sign`, `Verify` & `Debug` all unhide and show the correct dialog.

Testing https://github.com/bitcoin/bitcoin/pull/14123/commits/eed869b834a0d19318733420ed32f2f0df94d6be.
If `Command + H` or the menu item is used to hide the window:
- Clicking the task bar icon shows the window.
- Clicking `show/hide` does work.
- Click `show` (bottom of the menu ) does work.
- If `Send` or `Receive` is clicked, the window changes to the correct tab, and unhides.
- `Sign`, `Verify` & `Debug` all unhide and show the correct dialog. (this has also fixed the annoying "attention" bounce behaviour that was present before.)
- When the window has been minimised (CMD + M, which doesn't actually work at the moment..), only the sign, verify and debug items work, and they only display the sub window, the main window remains minimised. This seems to be a regression, as the options work with current master.

promag commented:
Thanks for the reviews! I'll fix it.

BTW, in macOS "show/hide" action could be removed since it's already there natively.

ken2812221 commented on src/qt/bitcoingui.cpp:
```diff
@@ -349,7 +349,9 @@ void BitcoinGUI::createActions()
+        connect(signMessageAction, &QAction::triggered, [this]{ showNormalIfMinimized(); });
         connect(signMessageAction, &QAction::triggered, [this]{ gotoSignMessageTab(); });
+        connect(verifyMessageAction, &QAction::triggered, [this]{ showNormalIfMinimized(); });
         connect(verifyMessageAction, &QAction::triggered, [this]{ gotoVerifyMessageTab(); });
```
I think this is better 
```
connect(signMessageAction, &QAction::triggered, [this]{ showNormalIfMinimized(); gotoSignMessageTab(); });
connect(verifyMessageAction, &QAction::triggered, [this]{ showNormalIfMinimized(); gotoSignMessageTab(); });

promag replied:
Keeping multiple connections for now.

gwillen replied:
I assume the other actions are only done that way for historical reasons -- prior to Qt5/C++11 this was not possible, and we only switched to using lambdas instead of the older syntax in 0.17 I believe.

ken2812221 commented on src/qt/bitcoingui.cpp:
```diff
@@ -1142,24 +1141,11 @@ void BitcoinGUI::showNormalIfMinimized(bool fToggleHidden)
+    if (!isHidden() && !isMinimized() && !GUIUtil::isObscured(this) && fToggleHidden) {
         hide();
+    } else {
+        GUIUtil::bringToFront(this);
```
If the window was minimized, it might need to call `showNormal`.

promag replied:
This should be fixed.

promag commented:
Rebased and added a fixup that `s/show/showNormal` after @ken2812221 https://github.com/bitcoin/bitcoin/pull/14123#discussion_r214781272.

promag commented on src/qt/guiutil.cpp:
```diff
@@ -62,6 +62,16 @@
+#if defined(Q_OS_MAC)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+// based on: https://github.com/Mozketo/LaunchAtLoginController/blob/master/LaunchAtLoginController.m
```
Sure, done.

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -357,6 +366,21 @@ bool isObscured(QWidget *w)
+
+    if (w) {
+        // activateWindow() (sometimes) helps with keyboard focus on Windows
+        w->activateWindow();
```
According to [Qt docs](https://doc.qt.io/qt-5/qwidget.html#activateWindow)

> Note that the window must be visible, otherwise activateWindow() has no effect.

Therefore, the safer sequence may be:
```
w->showNormal();  // or w->show();
w->activateWindow();
w->raise();

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -357,6 +366,21 @@ bool isObscured(QWidget *w)
+    if (w) {
+        // activateWindow() (sometimes) helps with keyboard focus on Windows
+        w->activateWindow();
+        w->showNormal();
```
Always `showNormal()` will break cases when then main window is maximized.

hebasto commented:
IMO `show()`/`showNormal()` should not be aggregated with `activateWindow()`->`raise()` sequence in a single function. Rationale: `show()` is reverse to `hide()`, but `showNormal()` restores the widget after it has been maximized or minimized.

jonasschnelli commented:
@promag: can you comment or address @hebasto's comments; rebase or close?

promag commented:
Rebased and addressed review comments.

hebasto commented:
tACK 56c61a3 

macOS 10.13.6 (17G65) + Qt 5.11.2 - works as expected, i.e. #13829 fixed

ken2812221 commented:
Tested ACK 56c61a33dd7f1bca4e5005654dd776ffe69c4c3b

promag reviewed (commented):
Rebased with #14597.

hebasto committed:
Remove obj_c for macOS Dock icon setting

Qt `setWindowIcon()` does this work.

hebasto committed:
Use Qt signal for macOS Dock icon click event

This moves the Dock icon click reaction code to the common place and
allows some cleanup in obj_c code.

According to the Apple's docs `class_replaceMethod` behaves as
`class_addMethod`, if the method identified by name does not yet exist;
or as `method_setImplementation`, if it does exist.

hebasto committed:
Remove obj_c for macOS Dock icon menu

Qt `setAsDockMenu()` does this work.

hebasto commented:
@promag 
The e0d1883d0e98c265db09b6a48b9c124f45edf898 commit you picked has a bug: https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-434257329
See: https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-434836662

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -60,6 +60,15 @@
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
+#include <objc/objc-runtime.h>
+#include <CoreFoundation/CoreFoundation.h>
```
This header is redundant.

ken2812221 reviewed (approved):
ACK d3becb175270d715daaffb75a037433a126a95f4

hebasto commented:
Tested d3becb175270d715daaffb75a037433a126a95f4:
- on Windows 10: bug #14222 fixed
- on macOS 10.13.6: bug #13829 fixed
There is a new bug on macOS: obscure main window with another app window, click Dock icon; main windows appears and hides. Investigating...
- on Linux Mint 19
@promag is it worth to combine a Linux-specific bugfix #14591 into this PR?

Also a tiny nit left: https://github.com/bitcoin/bitcoin/pull/14123#discussion_r230507680

Sjors commented:
re-tACK 0a656f8 on macOS 10.14.1 QT 5.11.2 (I didn't check the commits inherited from #14597)

Also +1 on doing Linux in another PR, it's already quite difficult to get any macOS changes through review.

hebasto commented:
re-tACK

Systems:
- Linux Mint 19 + Qt 5.9.5
- macOS 10.13.6 + Qt 5.11.2

Commits: 5796671e1dd8a2d0b1e750c2dce19a10af624095, 6fc21aca6d5e16c3ece104fec8e5b3df116893b4, 0a656f85a9c694f25b06c6464d6e986816eecd58
